File: app/src/main/java/me/zipi/navitotesla/service/NaviToTeslaService.java
package me.zipi.navitotesla.service;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;

import com.google.firebase.perf.metrics.AddTrace;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import me.zipi.navitotesla.AppRepository;
import me.zipi.navitotesla.BuildConfig;
import me.zipi.navitotesla.db.AppDatabase;
import me.zipi.navitotesla.db.PoiAddressEntity;
import me.zipi.navitotesla.exception.DuplicatePoiException;
import me.zipi.navitotesla.exception.ForbiddenException;
import me.zipi.navitotesla.exception.IgnorePoiException;
import me.zipi.navitotesla.exception.NotSupportedNaviException;
import me.zipi.navitotesla.model.ShareRequest;
import me.zipi.navitotesla.model.TeslaApiResponse;
import me.zipi.navitotesla.model.TeslaRefreshTokenRequest;
import me.zipi.navitotesla.model.Token;
import me.zipi.navitotesla.model.Vehicle;
import me.zipi.navitotesla.util.AnalysisUtil;
import me.zipi.navitotesla.util.PreferencesUtil;
import me.zipi.navitotesla.util.ResponseCloser;
import retrofit2.Response;


public class NaviToTeslaService {

    private final Context context;
    private final Pattern pattern = Pattern.compile("^(?:[가-힣]+\\s[가-힣]+[시군구]|세종시\\s[가-힣\\d]+[읍면동로])");
    AppRepository appRepository;

    public NaviToTeslaService(Context context) {
        this.context = context.getApplicationContext();
        appRepository = AppRepository.getInstance(this.context, AppDatabase.getInstance(this.context));
    }

    public boolean isAddress(String text) {
        return pattern.matcher(text).find();
    }

    private void makeToast(String text) {
        try {
            Log.i(this.getClass().getName(), text);
            AnalysisUtil.log(text);
            new Handler(Looper.getMainLooper()).post(() -> Toast.makeText(context, text, Toast.LENGTH_LONG).show());
        } catch (Exception e) {
            AnalysisUtil.recordException(e);
            e.printStackTrace();
        }

    }

    /**
     * 안내 종료
     */
    public void notificationClear() {
        PreferencesUtil.put(context, "lastAddress", "");
    }

    @AddTrace(name = "share")
    public void share(String packageName, String notificationTitle, String notificationText) {
        AnalysisUtil.setCustomKey("packageName", packageName);
        AnalysisUtil.setCustomKey("notificationTitle", notificationTitle);
        AnalysisUtil.setCustomKey("notificationText", notificationText);

        Bundle eventParam = new Bundle();
        eventParam.putString("package", packageName);

        try {

            String address = getAddress(packageName, notificationTitle, notificationText);
            String lastAddress = PreferencesUtil.getString(context, "lastAddress", "");

            if (!lastAddress.equals(address)) {
                try {
                    share(address);
                } catch (ForbiddenException e) {
                    AnalysisUtil.log("force expire token and retry...");
                    expireToken();
                    share(address);
                }
            } else {
                // 마지막 전송 주소와 동일
                // makeToast("목적지 전송 무시\n이전에 전송 요청한 주소와 동일함.");
                AnalysisUtil.logEvent("previous_request_address", eventParam);
            }
            appRepository.clearExpiredPoi();
        } catch (DuplicatePoiException e) {
            AnalysisUtil.logEvent("duplicated_address", eventParam);
            makeToast("목적지 전송 실패\n목적지 중복");
        } catch (NotSupportedNaviException e) {
            AnalysisUtil.logEvent("unsupported_navi", eventParam);
            AnalysisUtil.recordException(e);
            makeToast("목적지 전송 실패\n미지원 내비");
        } catch (IgnorePoiException e) {
            AnalysisUtil.logEvent("ignore_address", eventParam);
        } catch (ForbiddenException e) {
            makeToast("목적지 전송 실패\n인증 실패");
            AnalysisUtil.logEvent("error_share", eventParam);
            AnalysisUtil.recordException(e);
        } catch (Exception e) {
            Log.e(NaviToTeslaService.class.getName(), "thread inside error", e);
            makeToast("목적지 전송 실패\n내부 또는 API 오류");
            AnalysisUtil.logEvent("error_share", eventParam);
            AnalysisUtil.recordException(e);
            AnalysisUtil.sendUnsentReports();
        }
    }


    public void share(String address) throws IOException, ForbiddenException {
        PreferencesUtil.put(context, "lastAddress", address);
        if (address.length() > 0) {
            makeToast("목적지 전송 요청\n" + address);

            if (refreshToken() == null) {
                return;
            }
            Long id = loadVehicleId();
            if (BuildConfig.DEBUG) {
                makeToast("[DEBUG] 목적지 전송 Skip\n" + address);
                try {
                    Thread.sleep(500);
                } catch (InterruptedException ignore) {

                }
                return;
            }
            Response<TeslaApiResponse.ObjectType<TeslaApiResponse.Result>> response = appRepository.getTeslaApi().share(id, new ShareRequest(address)).execute();
            TeslaApiResponse.ObjectType<TeslaApiResponse.Result> result = null;
            if (response.isSuccessful()) {
                result = response.body();
            }
            if (result != null && result.getError() == null && result.getResponse() != null && result.getResponse().getResult()) {
                makeToast("목적지 전송 성공\n" + address);
                AnalysisUtil.log("send_success");
            } else {
                Log.w(NaviToTeslaService.class.getName(), response.toString());
                makeToast("목적지 전송 실패" + (result != null && result.getErrorDescription() != null ? "\n" + result.getErrorDescription() : ""));

                AnalysisUtil.log("send_fail");
                AnalysisUtil.setCustomKey("address", address);
                if (result != null && result.getErrorDescription() != null) {
                    AnalysisUtil.log("errorDescription: " + result.getErrorDescription());
                }
                if (!response.isSuccessful()) {
                    AnalysisUtil.log("Http response code: " + response.code());
                    if (response.errorBody() != null) {
                        AnalysisUtil.log("Http error response: " + response.errorBody().string());
                    }
                }

                RuntimeException exception;
                if (response.code() == 401) {
                    String errorString = "";
                    if (response.errorBody() != null) {
                        errorString = response.errorBody().string();
                    }
                    exception = new ForbiddenException(401, errorString);
                } else {
                    exception = new RuntimeException("Send address fail");
                }
                AnalysisUtil.recordException(exception);
                throw exception;
            }
            ResponseCloser.closeAll(response);
        }
    }

    @AddTrace(name = "getAddress")
    private String getAddress(String packageName, String notificationTitle, String notificationText)
            throws NotSupportedNaviException, DuplicatePoiException, IgnorePoiException, IOException {
        Bundle eventParam = new Bundle();
        eventParam.putString("package", packageName);

        PoiFinder poiFinder = PoiFinderFactory.getPoiFinder(packageName);

        String poiName = poiFinder.parseDestination(notificationText);
        if (poiName.length() == 0 || poiFinder.isIgnore(notificationTitle, notificationText)) {
            AnalysisUtil.logEvent("address_ignore_or_not_found", eventParam);
            throw new IgnorePoiException(packageName);
        }


        PoiAddressEntity poiAddressEntity = appRepository.getPoiSync(poiName);
        // 10 days cache
        String address;
        if (poiAddressEntity != null && !poiAddressEntity.isExpire()) {
            address = poiAddressEntity.getAddress();
            AnalysisUtil.logEvent("address_parse_cache", eventParam);
        } else if (isAddress(poiName)) {
            address = poiName;
            AnalysisUtil.logEvent("address_direct", eventParam);
        } else {
            address = poiFinder.findPoiAddress(poiName);
            AnalysisUtil.logEvent("address_parse_api", eventParam);
            appRepository.savePoi(poiName, address, false);
        }
        return address;
    }

    public Token getToken() {
        return PreferencesUtil.loadToken(context);
    }

    public void expireToken() {
        PreferencesUtil.expireToken(context);
    }

    @AddTrace(name = "refreshToken")
    public Token refreshToken() {
        return this.refreshToken(null);
    }

    public Token refreshToken(String refreshToken) {
        if (refreshToken == null) {
            Token token = PreferencesUtil.loadToken(context);
            if (token == null) {
                makeToast("Token 정보가 없습니다.");
                return null;
            } else if (!token.isExpire()) {
                return token;
            } else {
                refreshToken = token.getRefreshToken();
            }
        }

        AnalysisUtil.log("Start refresh access token");
        Token token = null;
        try {
            Response<Token> newToken = appRepository.getTeslaAuthApi().refreshAccessToken(new TeslaRefreshTokenRequest(refreshToken)).execute();

            if (newToken.isSuccessful() && newToken.body() != null) {
                PreferencesUtil.saveToken(context, newToken.body());
                token = newToken.body();
                AnalysisUtil.log("Success refresh access token");
            }
            ResponseCloser.closeAll(newToken);
        } catch (Exception e) {
            Log.w(this.getClass().getName(), "refresh token fail", e);
            makeToast("Token 갱신에 실패하였습니다.");
            AnalysisUtil.recordException(e);
        }
        if (token == null) {
            AnalysisUtil.log("fail refresh access token. token is null");
        }
        return token;
    }


    @AddTrace(name = "getVehicles")
    public List<Vehicle> getVehicles(Token token) {
        if (token == null) {
            token = PreferencesUtil.loadToken(context);
        }
        if (token == null || token.getRefreshToken() == null) {
            makeToast("Token 설정이 필요합니다.");
            return new ArrayList<>();
        }
        List<Vehicle> vehicles = new ArrayList<>();
        try {
            if (refreshToken() == null) {
                return vehicles;
            }
            Response<TeslaApiResponse.ListType<Vehicle>> response = appRepository.getTeslaApi().vehicles().execute();
            if (response.code() == 401) {
                makeToast("Token이 잘못되었습니다.");
            } else if (response.isSuccessful() && response.body() != null) {
                vehicles = response.body().getResponse();
            } else {
                Log.w(this.getClass().getName(), "get vehicle error: " + response.toString());
            }
            ResponseCloser.closeAll(response);

        } catch (Exception e) {
            Log.w(this.getClass().getName(), "get vehicle error", e);
            AnalysisUtil.recordException(e);
        }
        if (vehicles == null) {
            vehicles = new ArrayList<>();
        }
        if (vehicles.size() == 0) {
            makeToast("등록된 차량이 없습니다.");
        }
        return vehicles;
    }

    public void saveVehicleId(Long id) {
        PreferencesUtil.put(context, "vehicleId", id);
    }

    public Long loadVehicleId() {
        Long vid = PreferencesUtil.getLong(context, "vehicleId", 0L);
        return vid == null ? 0L : vid;
    }

    public void clearPoiCache() {
        appRepository.clearAllPoi();
    }
}

File: app/src/main/kotlin/me/zipi/navitotesla/service/NaviToTeslaService.kt
package me.zipi.navitotesla.service

import android.content.Context
import android.os.Bundle
import com.google.firebase.perf.metrics.AddTrace
import me.zipi.navitotesla.AppRepository
import me.zipi.navitotesla.R
import me.zipi.navitotesla.db.PoiAddressEntity
import me.zipi.navitotesla.exception.DuplicatePoiException
import me.zipi.navitotesla.exception.ForbiddenException
import me.zipi.navitotesla.exception.IgnorePoiException
import me.zipi.navitotesla.exception.NotSupportedNaviException
import me.zipi.navitotesla.model.Poi
import me.zipi.navitotesla.model.SendSettings
import me.zipi.navitotesla.model.ShareTransport
import me.zipi.navitotesla.model.TeslaApiResponse
import me.zipi.navitotesla.model.TeslaRefreshTokenRequest
import me.zipi.navitotesla.model.Token
import me.zipi.navitotesla.model.Vehicle
import me.zipi.navitotesla.service.place.DestinationAddressResolver
import me.zipi.navitotesla.service.poifinder.PoiFinderFactory
import me.zipi.navitotesla.service.share.SendPlanner
import me.zipi.navitotesla.service.share.TeslaShareByApi
import me.zipi.navitotesla.service.share.TeslaShareByApp
import me.zipi.navitotesla.ui.poi.PoiSelectionOverlay
import me.zipi.navitotesla.util.AnalysisUtil
import me.zipi.navitotesla.util.EnablerUtil
import me.zipi.navitotesla.util.PreferencesUtil
import me.zipi.navitotesla.util.RemoteConfigUtil
import me.zipi.navitotesla.util.ResponseCloser
import retrofit2.Response
import java.io.IOException
import java.util.Locale
import java.util.regex.Pattern

class NaviToTeslaService(
    context: Context,
) {
    private val context = context.applicationContext
    private val appRepository = AppRepository.getInstance()

    fun isAddress(text: String): Boolean = ADDRESS_PATTERN.matcher(text).find()

    private fun makeToast(
        text: String,
        level: AnalysisUtil.ToastLevel = AnalysisUtil.ToastLevel.INFO,
    ) = AnalysisUtil.makeToast(context = context, text = text, level = level)

    /**
     * 안내 종료
     */
    suspend fun notificationClear() {
        PreferencesUtil.put("lastAddress", "")
    }

    @AddTrace(name = "share")
    suspend fun share(
        packageName: String,
        notificationTitle: String?,
        notificationText: String?,
    ) {
        if (!EnablerUtil.isSendingCheck()) {
            AnalysisUtil.log("skip send share because condition")
            return
        }
        AnalysisUtil.setCustomKey("packageName", packageName)
        AnalysisUtil.setCustomKey("notificationTitle", notificationTitle ?: "")
        val eventParam = Bundle()
        eventParam.putString("package", packageName)
        try {
            appRepository.clearExpiredPoi()
            val poi = getPoi(packageName, notificationTitle, notificationText)
            val lastAddress = PreferencesUtil.getString("lastAddress", "")
            if (lastAddress != poi.getRoadAddress()) {
                try {
                    share(poi)
                } catch (_: ForbiddenException) {
                    AnalysisUtil.warn("force expire token and retry...")
                    expireToken()
                    share(poi)
                }
            } else {
                val deltaMs = if (lastShareAt > 0L) System.currentTimeMillis() - lastShareAt else -1L
                AnalysisUtil.log(
                    "skip share: duplicate dest, pkg=$packageName, addr=${poi.getRoadAddress()}, deltaMs=$deltaMs",
                )
                AnalysisUtil.logEvent("previous_request_address", eventParam)
            }
        } catch (e: DuplicatePoiException) {
            AnalysisUtil.logEvent("duplicated_address", eventParam)
            AnalysisUtil.log("duplicate poi name: " + e.poiName)
            val duplicatedToast = {
                makeToast(
                    text = context.getString(R.string.sendDestinationFail) + "\n" + context.getString(R.string.duplicatedPoiName),
                    level = AnalysisUtil.ToastLevel.WARN,
                )
            }
            val selectionEnabled = PreferencesUtil.getBoolean("duplicatePoiSelection", true)
            if (selectionEnabled && e.candidates.isNotEmpty()) {
                val shown =
                    PoiSelectionOverlay.show(context, e.candidates) {
                        AnalysisUtil.log("duplicate poi selection dismissed: " + e.poiName)
                        duplicatedToast()
                    }
                if (!shown) {
                    AnalysisUtil.warn("overlay permission denied for: " + e.poiName)
                    duplicatedToast()
                }
            } else {
                duplicatedToast()
            }
        } catch (e: NotSupportedNaviException) {
            AnalysisUtil.logEvent("unsupported_navi", eventParam)
            AnalysisUtil.recordException(e)
            makeToast(
                text = context.getString(R.string.sendDestinationFail) + "\n" + context.getString(R.string.unsupportedNavi),
                level = AnalysisUtil.ToastLevel.WARN,
            )
        } catch (_: IgnorePoiException) {
            AnalysisUtil.logEvent("ignore_address", eventParam)
        } catch (e: ForbiddenException) {
            makeToast(
                text = context.getString(R.string.sendDestinationFail) + "\n" + context.getString(R.string.authFail),
                level = AnalysisUtil.ToastLevel.WARN,
            )
            AnalysisUtil.logEvent("error_share", eventParam)
            AnalysisUtil.recordException(e)
        } catch (e: Exception) {
            AnalysisUtil.error("thread inside error", e)
            makeToast(
                text = context.getString(R.string.sendDestinationFail) + "\n" + context.getString(R.string.apiError),
                level = AnalysisUtil.ToastLevel.ERROR,
            )
            AnalysisUtil.logEvent("error_share", eventParam)
            AnalysisUtil.recordException(e)
            AnalysisUtil.sendUnsentReports()
        }
    }

    @Throws(IOException::class, ForbiddenException::class)
    suspend fun share(poi: Poi) {
        if (poi.isAddressEmpty()) {
            AnalysisUtil.warn("share skipped: empty poi name=${poi.poiName}, pkg=${poi.packageName}")
            PreferencesUtil.put(key = "lastAddress", value = "")
            makeToast(
                text = context.getString(R.string.sendDestinationFail) + "\n" + context.getString(R.string.addressNotFound),
                level = AnalysisUtil.ToastLevel.WARN,
            )
            return
        }

        // favorite/cache/좌표 무관 — classify 와 plan 이 각자 책임 안에서 처리.
        val searchability = DestinationAddressResolver.classify(poi)

        val shareMode = PreferencesUtil.getString("shareMode", "app")
        val transport =
            if (shareMode == "api" && PreferencesUtil.loadToken() != null) {
                ShareTransport.API
            } else {
                ShareTransport.APP
            }
        val settings =
            SendSettings(
                defaultMode = PreferencesUtil.getDefaultSendMode(),
                fallbackMode = PreferencesUtil.getFallbackSendMode(),
                treatUnknownAsNotSearchable = RemoteConfigUtil.getBoolean(RemoteConfigUtil.KEY_SEND_UNKNOWN_AS_NOT_SEARCHABLE),
                shareTransport = transport,
                locale = Locale.getDefault(),
            )

        val payload =
            SendPlanner.plan(
                poi = poi,
                searchability = searchability,
                isDuplicateSelected = poi.isDuplicate,
                settings = settings,
            )

        PreferencesUtil.put(key = "lastAddress", value = poi.getRoadAddress())

        lastShareAt = System.currentTimeMillis()
        makeToast(context.getString(R.string.requestSend) + "\n" + payload.displayText)
        if (transport == ShareTransport.API) {
            if (refreshToken() == null) {
                return
            }
            val id = loadVehicleId()
            TeslaShareByApi(context, id, poi.packageName).share(payload)
        } else {
            TeslaShareByApp(context, poi.packageName).share(payload)
        }
    }

    @AddTrace(name = "getAddress")
    @Throws(
        NotSupportedNaviException::class,
        DuplicatePoiException::class,
        IgnorePoiException::class,
        IOException::class,
    )
    private suspend fun getPoi(
        packageName: String,
        notificationTitle: String?,
        notificationText: String?,
    ): Poi {
        val eventParam = Bundle()
        eventParam.putString("package", packageName)
        val poiFinder = PoiFinderFactory.getPoiFinder(packageName)
        val poiName = poiFinder.parseDestination(notificationText ?: "")
        if (poiName.isEmpty() ||
            poiFinder.isIgnore(
                notificationTitle ?: "",
                notificationText ?: "",
            )
        ) {
            AnalysisUtil.logEvent("address_ignore_or_not_found", eventParam)
            throw IgnorePoiException(packageName)
        }
        val poiAddressEntity = appRepository.getPoiSync(poiName, packageName)
        val poi: Poi
        if (poiAddressEntity != null && !poiAddressEntity.isExpire) {
            // favorite 의 원래 packageName 보존 — cross-package favorite 의 cache key 일관성 유지.
            // 빈 packageName(글로벌 favorite) 인 경우만 share 출처 packageName 으로 보정.
            val effectivePackage = poiAddressEntity.packageName?.takeIf { it.isNotEmpty() } ?: packageName
            poi = poiAddressEntity.toPoi().copy(packageName = effectivePackage)
            AnalysisUtil.logEvent("address_parse_cache", eventParam)
        } else if (isAddress(poiName)) {
            poi =
                Poi(
                    poiName = poiName,
                    roadAddress = poiName,
                    address = null,
                    longitude = null,
                    latitude = null,
                    packageName = packageName,
                )
            appRepository.savePoi(poi, false)
            AnalysisUtil.logEvent("address_direct", eventParam)
        } else {
            poi = poiFinder.findPoi(poiName, packageName)
            AnalysisUtil.logEvent("address_parse_api", eventParam)
            appRepository.savePoi(poi, false)
        }
        return poi
    }

//    val token: Token?
//        get() = PreferencesUtil.loadToken()

    suspend fun getToken(): Token? = PreferencesUtil.loadToken()

    private suspend fun expireToken() {
        PreferencesUtil.expireToken()
    }

    @AddTrace(name = "refreshToken")
    suspend fun refreshToken(): Token? = this.refreshToken(null)

    suspend fun refreshToken(refreshToken: String?): Token? {
        var actualRefreshToken = refreshToken
        if (refreshToken == null) {
            val token = PreferencesUtil.loadToken()
            if (token == null) {
                makeToast(context.getString(R.string.notExistsToken), AnalysisUtil.ToastLevel.WARN)
                return null
            } else if (!token.isExpire()) {
                return token
            } else {
                actualRefreshToken = token.refreshToken
            }
        }
        AnalysisUtil.log("Start refresh access token")
        var token: Token? = null
        try {
            val newToken: Response<Token> =
                appRepository.teslaAuthApi.refreshAccessToken(
                    TeslaRefreshTokenRequest(actualRefreshToken),
                )
            if (newToken.isSuccessful) {
                newToken.body()?.let {
                    PreferencesUtil.saveToken(it)
                    token = it
                    AnalysisUtil.log("Success refresh access token")
                }
            }
            ResponseCloser.closeAll(newToken)
        } catch (e: Exception) {
            AnalysisUtil.warn("refresh token fail", e)
            makeToast(context.getString(R.string.refreshTokenError), AnalysisUtil.ToastLevel.WARN)
            AnalysisUtil.recordException(e)
        }
        if (token == null) {
            AnalysisUtil.warn("fail refresh access token. token is null")
        }
        return token
    }

    @AddTrace(name = "getVehicles")
    suspend fun getVehicles(token: Token?): List<Vehicle> {
        var actualToken = token
        if (token == null) {
            actualToken = PreferencesUtil.loadToken()
        }
        if (actualToken?.refreshToken == null) {
            makeToast(context.getString(R.string.requireToken), AnalysisUtil.ToastLevel.WARN)
            return emptyList()
        }
        val vehicles = mutableListOf<Vehicle>()
        try {
            if (refreshToken() == null) {
                return vehicles
            }
            val response: Response<TeslaApiResponse.ListType<Map<String, Any>>> = appRepository.teslaApi.products()
            if (response.code() == 401) {
                makeToast(context.getString(R.string.invalidToken), AnalysisUtil.ToastLevel.WARN)
            } else if (response.isSuccessful) {
                response
                    .body()
                    ?.response
                    ?.filter { it.containsKey("vin") && it.containsKey("vehicle_id") }
                    ?.map {
                        Vehicle(
                            id = (it["id"] as Number).toLong(),
                            vehicleId = (it["vehicle_id"] as Number).toLong(),
                            displayName = it["display_name"].toString(),
                            state = it["state"].toString(),
                        )
                    }?.let { vehicles.addAll(it) }
            } else {
                AnalysisUtil.warn("get vehicle error: $response")
            }
        } catch (e: Exception) {
            AnalysisUtil.warn("get vehicle error", e)
            AnalysisUtil.recordException(e)
        }
        if (vehicles.isEmpty()) {
            makeToast(text = context.getString(R.string.noVehicle), level = AnalysisUtil.ToastLevel.WARN)
        }
        return vehicles
    }

    suspend fun saveVehicleId(id: Long) {
        PreferencesUtil.put("vehicleId", id)
    }

    suspend fun loadVehicleId(): Long = PreferencesUtil.getLong("vehicleId", 0L)

    companion object {
        private val ADDRESS_PATTERN = Pattern.compile("^(?:[가-힣]+\\s[가-힣]+[시군구]|(?:세종시|세종특별시|세종특별자치시)\\s[가-힣\\d]+[읍면동로])\\s")

        @Volatile
        private var lastShareAt: Long = 0L
    }
}
